server/application/application.go

```diff
@@ -140,52 +141,98 @@ func NewServer(
 // getAppEnforceRBAC gets the Application with the given name in the given namespace. If no namespace is
 // specified, the Application is fetched from the default namespace (the one in which the API server is running).
 //
-// If the Application does not exist, then we have no way of determining if the user would have had access to get that
-// Application. Verifying access requires knowing the Application's name, namespace, and project. The user may specify,
-// at minimum, the Application name.
+// If the user does not provide a "project," then we have to be very careful how we respond. If an app with the given
+// name exists, and the user has access to that app in the app's project, we return the app. If the app exists but the
+// user does not have access, we return "permission denied." If the app does not exist, we return "permission denied" -
+// if we responded with a 404, then the user could infer that the app exists when they get "permission denied."
 //
-// So to prevent a malicious user from inferring the existence or absense of the Application or namespace, we respond
-// "permission denied" if the Application does not exist.
-func (s *Server) getAppEnforceRBAC(ctx context.Context, action, namespace, name string, getApp func() (*appv1.Application, error)) (*appv1.Application, error) {
+// If the user does provide a "project," we can respond more specifically. If the user does not have access to the given
+// app name in the given project, we return "permission denied." If the app exists, but the project is different from
+func (s *Server) getAppEnforceRBAC(ctx context.Context, action, project, namespace, name string, getApp func() (*appv1.Application, error)) (*appv1.Application, error) {
 	logCtx := log.WithFields(map[string]interface{}{
 		"application": name,
 		"namespace":   namespace,
 	})
+	if project != "" {
+		// The user has provided everything we need to perform an initial RBAC check.
+		givenRBACName := security.AppRBACName(s.ns, project, namespace, name)
+		if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplications, action, givenRBACName); err != nil {
+			logCtx.WithFields(map[string]interface{}{
+				"project":                project,
+				argocommon.SecurityField: argocommon.SecurityMedium,
+			}).Warnf("user tried to %s application which they do not have access to: %s", action, err)
+			// Do a GET on the app. This ensures that the timing of a "no access" response is the same as a "yes access,
+			// but the app is in a different project" response. We don't want the user inferring the existence of the
+			// app from response time.
+			_, _ = getApp()
```

Comment:
I agree to the point of not letting the user know of the existence of the app based on response times, but would it be wise to add a wait here instead of actually trying to get the app here? Wdyt?

Reply:
I think the problem with adding a wait is that we don't know the network conditions where Argo CD is running. So we don't know how long to wait.

Reply:
Fair point..

Reply:
Yep, we could do it on line 171. But then we'd need to set a boolean to determine whether we exit after `getApp` or proceed. I think this is a bit more readable, but I don't feel strongly either way.